src/Transports.AspNetCore/WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,253 @@
+            while (true)
+            {
+                var result = await _webSocket.ReceiveAsync(bufferMemory, RequestAborted);
+                if (result.MessageType == WebSocketMessageType.Close)
+                {
+                    // prevent any more messages from being queued
+                    operationMessageProcessor.Dispose();
+                    // send a close request if none was sent yet
+                    if (!_outputClosed.Task.IsCompleted)
+                    {
+                        // queue the closure
+                        _ = CloseAsync();
+                        // wait until the close has been sent
+                        var completedTask = Task.WhenAny(
+                            _outputClosed.Task,
+                            Task.Delay(_closeTimeoutMs, RequestAborted));
+                    }
+                    // quit
+                    return;
+                }
+                // if this is the last block terminating a message
+                if (result.EndOfMessage)
+                {
+                    // if only one block of data was sent for this message
+                    if (receiveStream.Length == 0)
+                    {
+                        // if the message is empty, skip to the next message
+                        if (result.Count == 0)
+                            continue;
+                        // read the message
+                        bufferStream.ResetLength(result.Count);
+                        var message = await _serializer.ReadAsync<OperationMessage>(bufferStream, RequestAborted);
+                        // dispatch the message
+                        if (message != null)
+                            await OnDispatchMessageAsync(operationMessageProcessor, message);
+                    }
+                    else
+                    {
+                        // if there is any data in this block, add it to the buffer
+                        if (result.Count > 0)
+                            receiveStream.Write(buffer, 0, result.Count);
+                        // read the message from the buffer
+                        receiveStream.Position = 0;
+                        var message = await _serializer.ReadAsync<OperationMessage>(receiveStream, RequestAborted);
+                        // clear the buffer
+                        receiveStream.SetLength(0);
+                        // dispatch the message
+                        if (message != null)
+                            await OnDispatchMessageAsync(operationMessageProcessor, message);
+                    }
+                }
+                else
+                {
+                    // if there is any data in this block, add it to the buffer
+                    if (result.Count > 0)
+                        receiveStream.Write(buffer, 0, result.Count);
+                }
+            }
+        }
+        catch (WebSocketException)
+        {
```

Comment:
Why empty cach?

Reply:
It often happens during normal closure and I did not find another way to prevent it.

Reply:
I'm unable to reproduce the problem now.  But practically, the only situation when a WebSocketException should occur is when the client disconnects in an incorrect manner or sends an invalid WebSocket packet, in which case it's not a server problem (and hence should not be logged).  Still, let's provide a protected method so that the exception could be handled if desired.

Reply:
Actually, inherited classes can wrap ExecuteAsync if they want to catch that exception.  They would also need to wrap HandleMessageAsync so as to catch exceptions while sending.

Reply:
I'm just changing this line to

```cs
catch (WebSocketException) when (RequestAborted.IsCancellationRequested)
```

Reply:
Yes, they can wrap but if we will keep exception swallowing here then no exception bubbles up.

Reply:
Ok…So now it’s only swallowed when it’s logical to do so - ok??

Reply:
My thought is that middleware and code in general should handle expected errors and only unexpected errors should fall through. In this manner user exception handling will catch unexpected errors. 

For example, when I write any TCP listener code, I always catch socket exceptions, because it is normal for a socket exception to occur when there is a disconnection. And that’s not a abnormal error. It’s typical. 

I’m not sure how the WebSocket code is written, but it certainly should not be throwing exceptions when the socket is disconnected, and probably no other WebSocketExceptions either.

Reply:
> So now it’s only swallowed when it’s logical to do so

Seems so. Most likely such an implementation is good.